### PR TITLE
Make sure units move to/stay on the appropriate movement layer after stopping

### PR DESCRIPTION
**Pull request by obrakmann**
Supersedes #15919, see the discussion there.

### CODE REVIEW

**Comment by MustaphaTR** [comment]
Units given stop command on a high-bridge, which below the bridge is also passable by the unit, get a move command to below the bridge, they no longer phase thru them but still (try to) move there. Same doesn't happen for Tunnel because move command is ignored.

**Comment by obrakmann** [comment]
Yup, it is. Fixed, and added a big-ass comment to explain all this mess.

**Comment by mazarf** [comment]
I revisited this and found IMO a simpler way to handle this.  Instead of shortening the movement path, this makes the unit stop then immediately gives it an order to leave the tunnel in OnBecomingIdle.

The tunnel exit Location is found in EntersTunnels.CustomLayerChanged.  This searches for the closest tunnel entrance and retrieves the exit from that.  I'm not sure if this is the best way to find the exit but this was what I came up with.

https://github.com/OpenRA/OpenRA/compare/bleed...mazarf:fix-tunnel-weirdness

**Comment by pchote** [comment]
Just a note that the `Move` changes here will be made redundant by #16246.  We'll need to test the rest of this once that has been merged.

**Comment by pchote** [comment]
#16246 has been merged, so this needs a rebase now. I think the `ICustomMovementLayer.PreventsStopping` is still probably worthwhile for decoupling `Move` from the specific layer implementations, even if the original bug has been fixed.

**Comment by obrakmann** [comment]
Rebased and updated. This now only fixes the remaining issue where units on high bridges would float to the ground after stopping.

File: OpenRA.Mods.Common/Traits/Mobile.cs [hunk 1]
```diff
@@ -669,7 +669,10 @@ void IDeathActorInitModifier.ModifyDeathActorInit(Actor self, TypeDictionary ini
 
 		void INotifyBecomingIdle.OnBecomingIdle(Actor self)
 		{
-			if (TopLeft.Layer == 0)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: If we keep the `self.Location.Layer == 0` check by itself at the start we can avoid the `TraitsImplementing` lookup in 99.9% of cases.

**Review by pchote — COMMENTED** [review]
Just one nit, otherwise LGTM

**Comment by obrakmann** [comment]
Updated

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/4b8471b6bd886e9d4e1abc7b79dafc5999ef62ef...f374d1e6994fdc89591652d33cf4414dfcc4ca81

**Pull request merged**